File: Source/Expression/solver_context.cpp
#include "solver_context.hpp"
#include <iostream>

namespace expression::solve {
  tree::Tree SimpleContext::reduce(tree::Tree input) {
    return context.reduce(input);
  }
  bool SimpleContext::needs_more_arguments(tree::Tree const& input) {
    for(auto const& rule : context.rules) {
      pattern::Tree const* pat = &rule.pattern;
      while(pattern::Apply const* app = pat->get_if_apply()) {
        if(term_matches(input, app->lhs)) {
          return true;
        }
        pat = &app->lhs;
      }
    }
    return false;
  }
  bool SimpleContext::is_head_closed(tree::Tree const& input, std::unordered_set<std::uint64_t> const& variables) {
    auto unfolding = unfold_ref(input);
    if(auto* ext = unfolding.head->get_if_external()) {
      return context.external_info[ext->index].is_axiom;
    } else {
      return true; //argument head
    }
  }
  void SimpleContext::add_rule(Rule rule, rule_explanation::Any) {
    context.rules.push_back(std::move(rule));
  }
  std::uint64_t SimpleContext::introduce_variable(introduction_explanation::Any) {
    auto ret = context.external_info.size();
    context.external_info.push_back({
      .is_axiom = false
    });
    return ret;
  }
  bool SimpleContext::term_depends_on(std::uint64_t term, std::uint64_t check) {
    return term == check;
  }

}

File: Source/Expression/solver_context.py
# Source Generator

shape = TypeErasedKind(namespace = "expression::solver", name = "Context", functions = [
    MemberFunction("simplify",
        ret="Simplification",
        args=[("tree::Expression","base")]
    ),
    MemberFunction("define_variable",
        ret="void",
        args=[("std::uint64_t","variable"),("std::uint64_t", "args"),("tree::Expression", "body")]
    ),
    #MemberFunction("introduce_variable",
    #    ret="std::uint64_t"
    #),
    MemberFunction("term_depends_on",
        ret="bool",
        args=[("std::uint64_t", "term"), ("std::uint64_t", "possible_dependency")]
    )
])

main_output = get_output("THIS_impl")

main_output.write(
    shape,
    relative_includes=["solver_context_interface.hpp"]
)
